mmdet3d/models/dense_heads/groupfree3d_head.py

```diff
@@ -0,0 +1,1047 @@
+
+            center_loss_sum += center_loss
+
+            # calculate direction class loss
+            dir_class_loss = self.dir_class_loss(
+                bbox_preds[f'dir_class{suffix}'].transpose(2, 1),
+                dir_class_targets,
+                weight=box_loss_weights)
+
+            dir_class_loss_sum += dir_class_loss
+
+            # calculate direction residual loss
+            heading_label_one_hot = size_class_targets.new_zeros(
+                (batch_size, proposal_num, self.num_dir_bins))
+            heading_label_one_hot.scatter_(2, dir_class_targets.unsqueeze(-1),
+                                           1)
+            dir_res_norm = torch.sum(
+                bbox_preds[f'dir_res_norm{suffix}'] * heading_label_one_hot,
+                -1)
+            dir_res_loss = self.dir_res_loss(
+                dir_res_norm, dir_res_targets, weight=box_loss_weights)
+
+            dir_res_loss_sum += dir_res_loss
+
+            if self.size_cls_agnostic:
+                # calculate class-agnostic size loss
+                size_reg_loss = self.size_reg_loss(
+                    bbox_preds[f'size{suffix}'],
+                    assigned_size_targets,
+                    weight=box_loss_weights_expand)
+
+                size_reg_loss_sum += size_reg_loss
+
+            else:
+                # calculate size class loss
+                size_class_loss = self.size_class_loss(
+                    bbox_preds[f'size_class{suffix}'].transpose(2, 1),
+                    size_class_targets,
+                    weight=box_loss_weights)
+
+                size_class_loss_sum += size_class_loss
+
+                # calculate size residual loss
+                one_hot_size_targets = size_class_targets.new_zeros(
+                    (batch_size, proposal_num, self.num_sizes))
+                one_hot_size_targets.scatter_(2,
+                                              size_class_targets.unsqueeze(-1),
+                                              1)
+                one_hot_size_targets_expand = one_hot_size_targets.unsqueeze(
+                    -1).expand(-1, -1, -1, 3).contiguous()
+                size_residual_norm = torch.sum(
+                    bbox_preds[f'size_res_norm{suffix}'] *
+                    one_hot_size_targets_expand, 2)
+                box_loss_weights_expand = box_loss_weights.unsqueeze(
+                    -1).expand(-1, -1, 3)
+                size_res_loss = self.size_res_loss(
+                    size_residual_norm,
+                    size_res_targets,
+                    weight=box_loss_weights_expand)
+
+                size_res_loss_sum += size_res_loss
+
+            # calculate semantic loss
+            semantic_loss = self.semantic_loss(
+                bbox_preds[f'sem_scores{suffix}'].transpose(2, 1),
+                mask_targets,
+                weight=box_loss_weights)
+
+            semantic_loss_sum += semantic_loss
+
+        objectness_loss_sum /= len(suffixes)
```

Comment:
we do not need 
```
objectness_loss_sum /= len(suffixes)
semantic_loss_sum /= len(suffixes)
center_loss_sum /= len(suffixes)
dir_class_loss_sum /= len(suffixes)
dir_res_loss_sum /= len(suffixes)
```
because sometimes we want to record and analyze the loss of each stage to better understand the training process. So we can simply put the losses of each stage to the loss dict and mark their stages like `s2.loss_semantic ` or `d2.loss_semantic`. See example in [DETR](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/dense_heads/detr_head.py)